### PR TITLE
[16.0] base_exception: Improves performances

### PR DESCRIPTION
**Pull request by lmignon**
To follow the changes you can read the commits one by one.

### CODE REVIEW

**Comment by lmignon** [comment]
ping @rousseldenis

**Review by rousseldenis — APPROVED** [review]
Code review.

Indeed, before :

image


After :

image



Even if few tests, quite 25% gain...

File: base_exception/models/exception_rule.py [hunk 1]
```diff
@@ -70,3 +70,59 @@ def _get_domain(self):
...
         return safe_eval(self.domain)
+
+    def _get_rules_info_for_domain(self, domain):
+        """returns the rules that match the domain
+
+        This method will call _get_cached_rules_for_domain to get the rules
+        that match the domain. It will then transform the result into a list
+        of objects that have the same attributes as the exception.rule model.
+        """
+        cache_info = self._get_cached_rules_for_domain(tuple(domain)).copy()
```
Comment (hparfr) [inline base_exception/models/exception_rule.py]: why copy() ?

File: base_exception/models/exception_rule.py [hunk 2]
```diff
@@ -70,3 +70,59 @@ def _get_domain(self):
...
+        return {
+            "id": self.id,
+            "name": self.name,
+            "description": self.description,
+            "sequence": self.sequence,
+            "model": self.model,
+            "exception_type": self.exception_type,
+            "domain": self._get_domain()
+            if self.exception_type == "by_domain"
+            else None,
```
Comment (hparfr) [inline base_exception/models/exception_rule.py:118]: is None a valid domain ?

**Review by hparfr — APPROVED** [review]
thanks for addressing this subject.

Code read; just two small remarks

File: base_exception/models/exception_rule.py (hunk 1 above)
Comment (lmignon) [reply base_exception/models/exception_rule.py]: should be removed... a trace from a previous implementation

File: base_exception/models/exception_rule.py (hunk 2 above)
Comment (lmignon) [reply base_exception/models/exception_rule.py:118]: No but the domain is only used  if `exception_type == "by_domain"`

**Review by hparfr — APPROVED** [review]
LGTM

**Comment by hparfr** [comment]
can we merge this one ?

**Comment by lmignon** [comment]
> can we merge this one ?

@hparfr yes plz

**Comment by hparfr** [comment]
I'm not PSC here :superhero:

**Pull request merged**